api/v1alpha2/awsmachinetemplate_webhook.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha2
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (r *AWSMachineTemplate) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha2-awsmachinetemplate,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=awsmachinetemplates,versions=v1alpha2,name=vawsmachinetemplate.kb.io
+
+var _ webhook.Validator = &AWSMachineTemplate{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachineTemplate) ValidateCreate() error {
+	// there's no validation for AWSMachineTempalate on create
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachineTemplate) ValidateUpdate(old runtime.Object) error {
+	oldAWSMachineTemplate := old.(*AWSMachineTemplate)
+	return validateUpdate(&r.Spec.Template.Spec, &oldAWSMachineTemplate.Spec.Template.Spec)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachineTemplate) ValidateDelete() error {
+	// there's no validation for AWSMachineTempalate on delete
+	return nil
+}
+
+// validateUpdate checks that no immutable fields have been updated
+func validateUpdate(spec *AWSMachineSpec, oldSpec *AWSMachineSpec) error {
```

Comment:
Are there any fields that we want to be mutable in the template types? @ncdc @vincepri I thought we had wanted to avoid allowing mutability of the template types and instead wanted to require users to update the reference in the MachineDeployment/MachineSet?

If we do want to force full immutability, then we can likely create a generic webhook in CAPI that blocks any update, and just register the webhook for the AWSMachineTemplate type here in CAPA.

Reply:
We can allow updates to metadata, but probably all other updates should be rejected. I think we could probably write some generic validation helpers in CAPI, but afaik we still need to do the full webhook code generation & registration for each type.

Reply:
According to the docs, we can configure which resources are subject to which admission webhooks at runtime: https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#configure-admission-webhooks-on-the-fly

Reply:
We should avoid naming changes to resources given that they're references elsewhere, unless the OwnerReferences get updated automagically?

Reply:
@ncdc I just want to avoid us requiring  all providers to need to re-implement the exact same thing

Reply:
I think that's fair, +1

Reply:
@detiber are you thinking that there would be one `ValidatingWebhookConfiguration` in CAPI and we would have documentation for how people installing providers would need to update it to include their resources in the generic template validator?

Reply:
@ncdc possibly, we could also potentially give RBAC permissions to providers to register the types in the ValidatingWebhookConfiguration themselves when the controller manager starts up.

Reply:
Can we dynamically watch unstructured objects in a group? And maybe discover those groups automatically?

CAPI should already have permissions for bootstrap.* and infrastructure.* groups

Reply:
We typically advise that registering things like CRDs and webhook configurations are tasks better left to admin accounts. I would be hesitant to let the controllers do this.

Reply:
This sounds like something that clusteradm could handle long-term

Reply:
Maybe we're talking about different things, I was thinking to make this Webhook server available by default in CAPI, and give it permissions to the groups and validate those objects

Reply:
@vincepri yes, we are talking about different things. Thanks for pointing that out. So you'd have this generic webhook use API discovery and look for `*Template` types? Although that wouldn't automatically cover third party providers whose API groups are different.

Reply:
As long as it's documented, I think that should be fine. Technically, the same applies for today's default RBAC permissions in CAPI, where we assume `bootstrap.<>` and `infrastructure.<>` as default

Reply:
@ncdc agreed that clusteradm would be the best place long term to handle this

Reply:
@vincepri @ncdc the validating webhook requests are initiated by the API Server, I'm not sure there is a way to use API discovery to register types to watch at runtime as we do with the controllers.

Reply:
@detiber that's right. So the only way to handle this is with the details in the ValidatingWebhookConfiguration.

What do you all think about putting the generic *Template validation as a helper function in cluster-api, but at least for now, having each provider implement their own webhooks for their own templates (using the generic function from cluster-api)?

Reply:
@ncdc I think that's fine as long as we track improving that experience later when clusteradm is available.